src/transformers/models/llama/modeling_llama.py

```diff
@@ -1265,17 +1265,18 @@ def prepare_inputs_for_generation(
 
         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
         if inputs_embeds is not None and cache_position[0] == 0:
-            model_inputs = {"inputs_embeds": inputs_embeds}
+            model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
         else:
-            model_inputs = {"input_ids": input_ids}
+            # The clone here is for the same reason as for `position_ids`.
+            model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
 
         if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
-            if inputs_embeds is not None:
-                batch_size, sequence_length = inputs_embeds.shape
-                device = inputs_embeds.device
+            if model_inputs["inputs_embeds"] is not None:
```

Comment:
I don't understand why we don't use `input_embeds` ?

Reply:
It's the same here: we used `inputs_embeds` to initialize the `model_inputs` dict and we never reuse it
The issue is that if a user generates with input_embeds, they persist, and `sequence_length` is taken as dimension -2 of inputs_embeds for the causal mask, which will cause a shape mismatch

Reply:
mmm yeah but it's a bit weird generate never gives both inputds or input embeds. It cool that we explicit it, we should just end up with 1 if else!

Reply:
Well it seems I ended up with inputs_embeds and a size-0 input_ids, but not None :thinking: so wanted to warrant this. For the sole if/else ah yes you're right, you mean like

```python
if isinstance(past_key_values, StaticCache) and attention_mask.ndim == 2:
    if inputs_embeds is not None and cache_position[0] == 0:
        model_inputs = {"inputs_embeds": inputs_embeds, "input_ids": None}
        batch_size, sequence_length, _ = model_inputs["inputs_embeds"].shape
        device = model_inputs["inputs_embeds"].device
    else:
        model_inputs = {"input_ids": input_ids.clone(memory_format=torch.contiguous_format), "inputs_embeds": None}
        batch_size, sequence_length = model_inputs["input_ids"].shape
        device = model_inputs["input_ids"].device
```

will update!

Reply:
ah no, that kills the model_inputs initialization - you had smth in mind for the only if/else?

Reply:
not sure no, just creating the model_inputs after the if else using input_embeds and input_id, not uregnet le'ts just merge this!

Reply:
alright! yes, we can do it later - merging, thanks!